src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,566 @@
+	if (*localRTECandidate == NULL)
+	{
+		return false;
+	}
+
+	if (list_length(rteToSubqueryConverterContext->distributedTableList) > 0)
+	{
+		*distributedRTECandidate = linitial(
+			rteToSubqueryConverterContext->distributedTableList);
+	}
+	return *distributedRTECandidate != NULL ||
+		   rteToSubqueryConverterContext->hasSubqueryRTE;
+}
+
+
+/*
+ * GetRTEToSubqueryConverterReferenceRelId returns the underlying relation id
+ * if it is a valid one.
+ */
+static Oid
+GetRTEToSubqueryConverterReferenceRelId(
+	RTEToSubqueryConverterReference *rteToSubqueryConverterReference)
+{
+	if (rteToSubqueryConverterReference &&
+		rteToSubqueryConverterReference->rangeTableEntry)
+	{
+		return rteToSubqueryConverterReference->rangeTableEntry->relid;
+	}
+	return InvalidOid;
+}
+
+
+/*
+ * RemoveFromRTEToSubqueryConverterContext removes an element from
+ * the relevant list based on the relation id.
+ */
+static void
+RemoveFromRTEToSubqueryConverterContext(
+	RTEToSubqueryConverterContext *rteToSubqueryConverterContext, Oid relationId)
+{
+	if (IsLocalOrCitusLocalTable(relationId))
+	{
+		rteToSubqueryConverterContext->localTableList =
+			list_delete_first(rteToSubqueryConverterContext->localTableList);
+	}
+	else
+	{
+		rteToSubqueryConverterContext->distributedTableList =
+			list_delete_first(rteToSubqueryConverterContext->distributedTableList);
+	}
+}
+
+
+/*
+ * ShouldConvertLocalTableJoinsToSubqueries returns true if we should
+ * convert local-dist table joins to subqueries.
+ */
+static bool
+ShouldConvertLocalTableJoinsToSubqueries(List *rangeTableList, Oid resultRelationId)
+{
+	if (LocalTableJoinPolicy == LOCAL_JOIN_POLICY_NEVER)
+	{
+		/* user doesn't want Citus to enable local table joins */
+		return false;
+	}
+	if (!ContainsTableToBeConvertedToSubquery(rangeTableList, resultRelationId))
+	{
+		return false;
+	}
+	return true;
+}
+
+
+/*
+ * ShouldConvertDistributedTable returns true if we should convert the
+ * distributed table rte to a subquery. This will be the case if the distributed
+ * table has a unique index on a column that appears in filter.
+ */
+static bool
+ShouldConvertDistributedTable(FromExpr *joinTree,
+							  RTEToSubqueryConverterReference *
+							  rteToSubqueryConverterReference)
+{
+	if (rteToSubqueryConverterReference == NULL)
+	{
+		return false;
+	}
+	List *distRTEEqualityQuals =
+		FetchEqualityAttrNumsForRTEFromQuals(joinTree->quals,
+											 rteToSubqueryConverterReference->rteIndex);
+
+	Node *join = NULL;
+	foreach_ptr(join, joinTree->fromlist)
+	{
+		if (IsA(join, JoinExpr))
+		{
+			JoinExpr *joinExpr = (JoinExpr *) join;
+			distRTEEqualityQuals = list_concat(distRTEEqualityQuals,
+											   FetchEqualityAttrNumsForRTEFromQuals(
+												   joinExpr->quals,
+												   rteToSubqueryConverterReference->
+												   rteIndex)
+											   );
+		}
+	}
+
+	bool hasUniqueFilter = HasUniqueFilter(
+		rteToSubqueryConverterReference->rangeTableEntry,
+		rteToSubqueryConverterReference->
```

Comment:
there are a few of these "fat" function calls that would become more readable by using some local variables